File: sdk-typescript/src/messaging.ts
/**
 * Messaging and invocation functionality for Nexus SDK.
 */

import { NexusClient } from './client';

export interface Message {
  id: string;
  from_agent_id: string;
  to_agent_id: string;
  subject?: string;
  content: Record<string, unknown>;
  reply_to_id?: string;
  status: string;
  created_at: string;
  read_at?: string;
}

export interface MessageList {
  messages: Message[];
  total: number;
}

export interface Invocation {
  id: string;
  caller_agent_id: string;
  target_agent_id: string;
  capability_id: string;
  capability_name?: string;
  status: 'pending' | 'processing' | 'completed' | 'failed' | 'timeout';
  input_data: Record<string, unknown>;
  output_data?: Record<string, unknown>;
  error_message?: string;
  created_at: string;
  started_at?: string;
  completed_at?: string;
}

export interface InvocationList {
  invocations: Invocation[];
  total: number;
}

export interface PendingInvocation {
  id: string;
  caller_agent_id: string;
  caller_agent_name?: string;
  capability_name: string;
  input_data: Record<string, unknown>;
  created_at: string;
  timeout_seconds: number;
}

export interface PendingWork {
  invocations: PendingInvocation[];
  messages: Message[];
}

export interface WebhookConfig {
  endpoint_url: string;
  events: string[];
  active: boolean;
}

/**
 * Client for agent-to-agent messaging.
 */
export class MessagingClient {
  constructor(private client: NexusClient) {}

  /**
   * Send a message to another agent.
   */
  async send(
    toAgentId: string,
    content: Record<string, unknown>,
    options?: {
      subject?: string;
      replyToId?: string;
    }
  ): Promise<Message> {
    const data: Record<string, unknown> = {
      to_agent_id: toAgentId,
      content,
    };
    if (options?.subject) data.subject = options.subject;
    if (options?.replyToId) data.reply_to_id = options.replyToId;

    return this.client.post<Message>('/messages', data);
  }

  /**
   * Get messages received by this agent.
   */
  async inbox(options?: {
    unreadOnly?: boolean;
    limit?: number;
    offset?: number;
  }): Promise<MessageList> {
    const params = new URLSearchParams({
      inbox: 'true',
      unread_only: String(options?.unreadOnly ?? false),
      limit: String(options?.limit ?? 50),
      offset: String(options?.offset ?? 0),
    });
    return this.client.get<MessageList>(`/messages?${params}`);
  }

  /**
   * Get messages sent by this agent.
   */
  async sent(options?: { limit?: number; offset?: number }): Promise<MessageList> {
    const params = new URLSearchParams({
      inbox: 'false',
      limit: String(options?.limit ?? 50),
      offset: String(options?.offset ?? 0),
    });
    return this.client.get<MessageList>(`/messages?${params}`);
  }

  /**
   * Mark a message as read.
   */
  async markRead(messageId: string): Promise<Message> {
    return this.client.post<Message>(`/messages/${messageId}/read`, {});
  }
}

/**
 * Client for invoking capabilities on other agents.
 */
export class InvocationClient {
  constructor(private client: NexusClient) {}

  /**
   * Invoke a capability on another agent.
   */
  async invoke(
    agentId: string,
    capability: string,
    options?: {
      input?: Record<string, unknown>;
      timeoutSeconds?: number;
      asyncMode?: boolean;
    }
  ): Promise<Invocation> {
    const data = {
      input: options?.input ?? {},
      timeout_seconds: options?.timeoutSeconds ?? 30,
      async_mode: options?.asyncMode ?? false,
    };
    return this.client.post<Invocation>(`/invoke/${agentId}/${capability}`, data);
  }

  /**
   * Get an invocation by ID.
   */
  async get(invocationId: string): Promise<Invocation> {
    return this.client.get<Invocation>(`/invocations/${invocationId}`);
  }

  /**
   * List invocations for this agent.
   */
  async list(options?: {
    asCaller?: boolean;
    status?: string;
    limit?: number;
  }): Promise<InvocationList> {
    const params = new URLSearchParams({
      as_caller: String(options?.asCaller ?? true),
      limit: String(options?.limit ?? 50),
    });
    if (options?.status) params.set('status', options.status);
    return this.client.get<InvocationList>(`/invocations?${params}`);
  }

  /**
   * Get pending work for this agent (invocations and messages).
   */
  async pending(): Promise<PendingWork> {
    return this.client.get<PendingWork>('/agents/me/pending');
  }

  /**
   * Complete an invocation with a result.
   */
  async complete(
    invocationId: string,
    result: {
      output?: Record<string, unknown>;
      error?: string;
    }
  ): Promise<Invocation> {
    const data: Record<string, unknown> = {};
    if (result.output !== undefined) data.output = result.output;
    if (result.error !== undefined) data.error = result.error;
    return this.client.post<Invocation>(`/invocations/${invocationId}/complete`, data);
  }
}

/**
 * Client for managing webhooks.
 */
export class WebhookClient {
  constructor(private client: NexusClient) {}

  /**
   * Configure webhook for receiving invocations and messages.
   */
  async set(
    endpointUrl: string,
    events?: string[]
  ): Promise<WebhookConfig> {
    return this.client.post<WebhookConfig>('/agents/me/webhook', {
      endpoint_url: endpointUrl,
      events: events ?? ['invocation', 'message'],
    });
  }

  /**
   * Get current webhook configuration.
   */
  async get(): Promise<WebhookConfig | null> {
    return this.client.get<WebhookConfig | null>('/agents/me/webhook');
  }

  /**
   * Remove webhook configuration.
   */
  async remove(): Promise<void> {
    await this.client.delete('/agents/me/webhook');
  }
}

File: sdk-python/nexus_sdk/messaging.py
"""Messaging and invocation functionality for Nexus SDK."""

from typing import Any
from uuid import UUID


class MessagingClient:
    """Client for agent-to-agent messaging."""

    def __init__(self, http_client):
        self._http = http_client

    def send(
        self,
        to_agent_id: str | UUID,
        content: dict[str, Any],
        subject: str | None = None,
        reply_to_id: str | UUID | None = None,
    ) -> dict:
        """
        Send a message to another agent.

        Args:
            to_agent_id: Target agent's ID
            content: Message content (dict)
            subject: Optional message subject
            reply_to_id: Optional ID of message being replied to

        Returns:
            Message object with id, status, etc.
        """
        data = {
            "to_agent_id": str(to_agent_id),
            "content": content,
        }
        if subject:
            data["subject"] = subject
        if reply_to_id:
            data["reply_to_id"] = str(reply_to_id)

        return self._http.post("/messages", data)

    def inbox(
        self,
        unread_only: bool = False,
        limit: int = 50,
        offset: int = 0,
    ) -> dict:
        """
        Get messages received by this agent.

        Args:
            unread_only: Only return unread messages
            limit: Max messages to return
            offset: Pagination offset

        Returns:
            Dict with 'messages' list and 'total' count
        """
        params = {
            "inbox": "true",
            "unread_only": str(unread_only).lower(),
            "limit": limit,
            "offset": offset,
        }
        return self._http.get("/messages", params=params)

    def sent(self, limit: int = 50, offset: int = 0) -> dict:
        """
        Get messages sent by this agent.

        Args:
            limit: Max messages to return
            offset: Pagination offset

        Returns:
            Dict with 'messages' list and 'total' count
        """
        params = {
            "inbox": "false",
            "limit": limit,
            "offset": offset,
        }
        return self._http.get("/messages", params=params)

    def mark_read(self, message_id: str | UUID) -> dict:
        """
        Mark a message as read.

        Args:
            message_id: ID of the message

        Returns:
            Updated message object
        """
        return self._http.post(f"/messages/{message_id}/read", {})


class InvocationClient:
    """Client for invoking capabilities on other agents."""

    def __init__(self, http_client):
        self._http = http_client

    def invoke(
        self,
        agent_id: str | UUID,
        capability: str,
        input_data: dict[str, Any] | None = None,
        timeout_seconds: int = 30,
        async_mode: bool = False,
    ) -> dict:
        """
        Invoke a capability on another agent.

        Args:
            agent_id: Target agent's ID
            capability: Name of the capability to invoke
            input_data: Input data for the capability
            timeout_seconds: Timeout in seconds (1-300)
            async_mode: If True, return immediately with invocation ID

        Returns:
            Invocation object with id, status, output_data, etc.
        """
        data = {
            "input": input_data or {},
            "timeout_seconds": timeout_seconds,
            "async_mode": async_mode,
        }
        return self._http.post(f"/invoke/{agent_id}/{capability}", data)

    def get(self, invocation_id: str | UUID) -> dict:
        """
        Get an invocation by ID.

        Args:
            invocation_id: ID of the invocation

        Returns:
            Invocation object
        """
        return self._http.get(f"/invocations/{invocation_id}")

    def list(
        self,
        as_caller: bool = True,
        status: str | None = None,
        limit: int = 50,
    ) -> dict:
        """
        List invocations for this agent.

        Args:
            as_caller: If True, list invocations made by this agent.
                      If False, list invocations received.
            status: Filter by status (pending, processing, completed, failed)
            limit: Max invocations to return

        Returns:
            Dict with 'invocations' list and 'total' count
        """
        params = {
            "as_caller": str(as_caller).lower(),
            "limit": limit,
        }
        if status:
            params["status"] = status
        return self._http.get("/invocations", params=params)

    def pending(self) -> dict:
        """
        Get pending work for this agent (invocations and messages).

        Returns:
            Dict with 'invocations' and 'messages' lists
        """
        return self._http.get("/agents/me/pending")

    def complete(
        self,
        invocation_id: str | UUID,
        output: dict[str, Any] | None = None,
        error: str | None = None,
    ) -> dict:
        """
        Complete an invocation with a result.

        Args:
            invocation_id: ID of the invocation to complete
            output: Output data (if successful)
            error: Error message (if failed)

        Returns:
            Updated invocation object
        """
        data = {}
        if output is not None:
            data["output"] = output
        if error is not None:
            data["error"] = error
        return self._http.post(f"/invocations/{invocation_id}/complete", data)


class WebhookClient:
    """Client for managing webhooks."""

    def __init__(self, http_client):
        self._http = http_client

    def set(
        self,
        endpoint_url: str,
        events: list[str] | None = None,
    ) -> dict:
        """
        Configure webhook for receiving invocations and messages.

        Args:
            endpoint_url: URL to receive webhook calls
            events: List of events to receive (invocation, message)

        Returns:
            Webhook configuration
        """
        data = {
            "endpoint_url": endpoint_url,
            "events": events or ["invocation", "message"],
        }
        return self._http.post("/agents/me/webhook", data)

    def get(self) -> dict | None:
        """
        Get current webhook configuration.

        Returns:
            Webhook config or None if not set
        """
        return self._http.get("/agents/me/webhook")

    def remove(self) -> None:
        """Remove webhook configuration."""
        self._http.delete("/agents/me/webhook")
